crates/federate/src/worker.rs

```diff
@@ -67,60 +77,128 @@ pub(crate) struct InstanceWorker {
+    let pool = config.to_request_data().inner_pool().clone();
+    let state = FederationQueueState::load(&mut DbPool::Pool(&pool), instance.id).await?;
     let mut worker = InstanceWorker {
       instance,
       site_loaded: false,
       site: None,
       followed_communities: HashMap::new(),
       stop,
-      context,
+      config,
       stats_sender,
       last_full_communities_fetch: Utc.timestamp_nanos(0),
       last_incremental_communities_fetch: Utc.timestamp_nanos(0),
       state,
       last_state_insert: Utc.timestamp_nanos(0),
+      pool,
     };
-    worker.loop_until_stopped(pool).await
+    worker.loop_until_stopped().await
   }
   /// loop fetch new activities from db and send them to the inboxes of the given instances
   /// this worker only returns if (a) there is an internal error or (b) the cancellation token is cancelled (graceful exit)
-  pub(crate) async fn loop_until_stopped(
-    &mut self,
-    pool: &mut DbPool<'_>,
-  ) -> Result<(), anyhow::Error> {
-    let save_state_every = chrono::Duration::from_std(SAVE_STATE_EVERY_TIME).expect("not negative");
-
-    self.update_communities(pool).await?;
+  async fn loop_until_stopped(&mut self) -> Result<()> {
     self.initial_fail_sleep().await?;
+    let mut latest_id = self.get_latest_id().await?;
+
+    // activities that have been successfully sent but
+    // that are not the lowest number and thus can't be written to the database yet
+    let mut successfuls = BinaryHeap::<SendSuccessInfo>::new();
+    // number of activities that currently have a task spawned to send it
+    let mut in_flight: i64 = 0;
+
+    // each HTTP send will report back to this channel concurrently
+    let (report_send_result, mut receive_send_result) =
+      tokio::sync::mpsc::unbounded_channel::<SendActivityResult>();
     while !self.stop.is_cancelled() {
-      self.loop_batch(pool).await?;
-      if self.stop.is_cancelled() {
-        break;
-      }
-      if (Utc::now() - self.last_state_insert) > save_state_every {
-        self.save_and_send_state(pool).await?;
+      // check if we need to wait for a send to finish before sending the next one
+      // we wait if (a) the last request failed, only if a request is already in flight (not at the start of the loop)
+      // or (b) if we have too many successfuls in memory or (c) if we have too many in flight
+      let need_wait_for_event = (in_flight != 0 && self.state.fail_count > 0)
+        || successfuls.len() >= MAX_SUCCESSFULS
+        || in_flight >= *CONCURRENT_SENDS;
+      if need_wait_for_event || receive_send_result.len() > 4 {
+        // if len() > 0 then this does not block and allows us to write to db more often
+        // if len is 0 then this means we wait for something to change our above conditions,
+        // which can only happen by an event sent into the channel
+        self
+          .handle_send_results(&mut receive_send_result, &mut successfuls, &mut in_flight)
+          .await?;
```

Comment:
I would launch a separate background task for this instead of manually checking here. Can also potentially be moved into a separate struct/file.

Reply:
_maybe_, that's what I originally wanted to do, but the problem then is that you need some way for the "read activities loop" to pause and unpause depending on the state of the results handling tasks. That would require either another arbitrary sleep() or another channel for sending around values.

I intentionally put that inline in here because then the "read activities" loop knows exactly when to continue and nothing need any thread synchronization. Nothing in this struct needs atomics or locks because there's only a single task accessing anything, the only thing running in parallel is the actual sends.

I agree that passing muts to local variables is pretty... weird, I originally just had this part inlined, i moved it out so it's easier to grok the whole loop.

Reply:
> [recv_many](https://docs.rs/tokio/latest/tokio/sync/mpsc/struct.UnboundedReceiver.html#method.recv_many) blocks automatically if there are no items

Yes but the problem is that you're saying you want to have the recv be in a separate thread from the thread that queues new sends. So then how would the sending-queuer know when it's supposed to send new items? It's inline in here exactly because it blocks if there's no items and thus pauses the send loop for the exactly correct time until a new item should be sent.  The alternatives are hacky and would require more arbitrarily chosen delays. And this event is expected to happen 10+ times per second per instance so waiting would basically be a busy loop (sleep 10ms, recheck, sleep 10ms, recheck, ...)

Reply:
Makes sense. What about using  only a single `unbounded_channel` for activity send completion and `print_stats()`?

Reply:
_maybe_, it would work (I don't remember the details right now), but at least currently how it works is that 

```
SendManager --(has many)--> InstanceWorker --(has many)--> SendRetryTask

-----------create worker ->                ---------------> send activity
                                                              vvvv
                                                            fail or success
                                            <---report result--
                    <----order and aggrate results---
        <---send stats---
filter and print stats
```

So basically the stats channel is between sendmanager and instanceworker, but the send_result channel is between InstanceWorker and SendRetryTask. So the parties are different and incompatible and the data sent is also different.

For example, an out of order send result interests the instance worker but doesn't actually update anything in the DB or the stats because the worker needs to wait for older events first.